src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala

```diff
@@ -34,100 +38,165 @@ object StructuralVariant {
+      // Following DELLY, we use a threshold of 5 median absolute deviations above the median.
+      val maxNormalInsertSize = (insertStats.median + 5 * insertStats.mad).toInt
+
+      val exceptionalReads = readsInRange.filter(_.inferredInsertSize > maxNormalInsertSize)
+
+      ExceptionalReadsReturnType(
+        readsInRange,
+        insertSizes,
+        insertStats,
+        maxNormalInsertSize,
+        exceptionalReads
+      )
+    }
+
+    // Given two pairs of reads, is there a deletion which would make both of them have normal insert sizes?
+    // TODO: this assumes that both reads have read.start.pos < read.mate.pos
+    def areReadsIncompatible(read1: PairedMappedRead, read2: PairedMappedRead, maxNormalInsertSize: Long): Boolean = {
+      if (read1.read.start < read2.read.start) return areReadsIncompatible(read2, read1, maxNormalInsertSize)
+
+      // This matches the DELLY logic; the last five lines are copy/paste. See
+      // https://github.com/tobiasrausch/delly/blob/7464cacfe1d420ac5bd7ed40a3093a80a93a6964/src/tags.h#L441-L445
+      val pair1Min = Math.min(read1.read.start, read1.mate.start)
+      val pair1Max = pair1Min + read1.inferredInsertSize
+      val pair1ReadLength = read1.read.sequence.length
+
+      val pair2Min = Math.min(read2.read.start, read2.mate.start)
+      val pair2Max = pair2Min + read2.inferredInsertSize
+      val pair2ReadLength = read2.read.sequence.length
+
+      if ((pair2Min + pair2ReadLength - pair1Min) > maxNormalInsertSize) return true;
+      if ((pair2Max < pair1Max) && ((pair1Max + pair1ReadLength - pair2Max) > maxNormalInsertSize)) return true;
+      if ((pair2Max >= pair1Max) && ((pair2Max + pair2ReadLength - pair1Max) > maxNormalInsertSize)) return true;
+      if ((pair1Max < pair2Min) || (pair2Max < pair1Min)) return true;
+      return false;
+    }
+
+    // Given a list of reads with abnormally large inserts, construct a graph representation of the structural variants
+    // where:
+    // Vertices = paired reads
+    // Edges = two sets of paired reads could represent the same structural variant
+    // For more details, see the DELLY paper.
+    def buildVariantGraph(exceptionalReads: Iterable[PairedMappedRead], maxNormalInsertSize: Int): PairedReadGraph = {
+      val reads = exceptionalReads.toArray.sortBy(r => Math.min(r.read.start, r.mate.start))
+      val graph = MutableGraph[PairedMappedRead, WUnDiEdge]()
+
+      for {
+        i <- 0 until reads.length
+      } {
+        val read = reads(i)
+        val start = Math.min(read.read.start, read.mate.start)
+        var j = i + 1
+        var done = false
+        // Loop through subsequent reads until it's impossible to find another compatible one.
+        while (j < reads.length && !done) {
+          val nextRead = reads(j)
+          val nextStart = Math.min(nextRead.read.start, nextRead.mate.start)
+          if (Math.abs(nextStart + nextRead.read.sequence.length - start) > maxNormalInsertSize) {
```

Comment:
what does `nextStart + nextRead.read.sequence.length` mean if `nextStart` is actually the mate's start pos?